onnxruntime/core/providers/cuda/math/inverse.cc

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/cuda/cuda_common.h"
+#include "core/providers/cuda/math/unary_elementwise_ops_impl.h"
+
+namespace onnxruntime {
+namespace cuda {
+
+class Inverse final : public CudaKernel {
+ public:
+  explicit Inverse(const OpKernelInfo& info) : CudaKernel{info} {
+  }
+
+  Status ComputeInternal(OpKernelContext* context) const override;
+
+ private:
+  using Base = CudaKernel;
+  using CublasHandle = cublasHandle_t;
+
+  template <typename T>
+  struct ComputeImpl;
+};
+
+ONNX_OPERATOR_KERNEL_EX(
+    Inverse,
+    kOnnxDomain,
+    12,
+    kCudaExecutionProvider,
+    KernelDefBuilder()
+        .TypeConstraint("T", BuildKernelDefConstraints<float, double, MLFloat16>()),
+    Inverse);
+
+template <typename T>
+Status ComputeMatrixOffsets(T* workspace_data, size_t num_batches, size_t rows, IAllocatorUniquePtr<T*>& matrix_ptrs) {
+  std::vector<T*> cuda_ptrs;
+  const size_t matrix_size = rows * rows;
+  for (size_t i = 0; i < num_batches; ++i) {
+    cuda_ptrs.push_back(workspace_data);
+    workspace_data += matrix_size;
+  }
+  CUDA_RETURN_IF_ERROR(cudaMemcpy(matrix_ptrs.get(), cuda_ptrs.data(), sizeof(T*) * num_batches,
+                                  cudaMemcpyHostToDevice));
+  return Status::OK();
+}
+
+Status CheckForSingularity(const IAllocatorUniquePtr<int>& info, const std::unique_ptr<int[]>& info_cpu, size_t num_batches) {
+  // Let's check if any of the info values is non-zero
+  CUDA_RETURN_IF_ERROR(cudaMemcpy(info_cpu.get(), info.get(), sizeof(int) * num_batches,
+                                  cudaMemcpyDeviceToHost));
+  for (auto i = 0; i < num_batches; ++i) {
+    if (info_cpu[i] != 0) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Matrix is singular at batch:", i);
```

Comment:
Curious: For the CPU version, Does the Eigen implementation provide such a friendly message stating that the matrix is singular ?

Reply:
No, it does not. I would have to do Full Pivoting LU decomposition to provide such checks. And the standard provides flexibility in checking. Ke suggested I try a different approach for CUDA so the code may still change.

---
In reply to: [406930453](https://github.com/microsoft/onnxruntime/pull/3485#discussion_r406930453) [](ancestors = 406930453)